src/Microsoft.Extensions.DependencyInjection/ServiceProvider.cs

```diff
@@ -17,18 +17,19 @@ namespace Microsoft.Extensions.DependencyInjection
     /// </summary>
     internal class ServiceProvider : IServiceProvider, IDisposable
     {
-        private readonly ServiceProvider _root;
         private readonly ServiceTable _table;
         private bool _disposeCalled;
-
-        private readonly Dictionary<IService, object> _resolvedServices = new Dictionary<IService, object>();
         private List<IDisposable> _transientDisposables;
 
+        internal ServiceProvider Root { get; }
+        internal Dictionary<object, object> ResolvedServices { get; } = new Dictionary<object, object>();
+
         private static readonly Func<Type, ServiceProvider, Func<ServiceProvider, object>> _createServiceAccessor = CreateServiceAccessor;
+        private static CallSiteRuntimeResolver _callSiteRuntimeResolver = new CallSiteRuntimeResolver();
```

Comment:
Add a comment saying that this is completely stateless so it's safe to cache (unlike the expression generator)